### PR TITLE
Add Livy server to Spark

### PR DESCRIPTION
**Pull request by galgalesh**
This PR incorporates Andrew's https://github.com/juju-solutions/layer-apache-spark/pull/7 and the following changes:
- Run Livy as Hue user
- Use correct log directory for spark
- Separate hadoop-client, Spark and Livy `dist` settings in `layer.yaml`
- (mostly) separate Spark and Livy handlers
- Remove use of `kv` to store state in Livy and Spark class.

I'm not sure why the Livy `dist` config was under `hadoop-client`. It took me a while to understand what was really happening, so I just separated it to make it clearer.

### CODE REVIEW

**Comment by galgalesh** [comment]
I tested this with the Hue Charm. Everything seems to be working. However, I noticed that Hue/Livy maintains a constant YARN job that block one slave from being used by other jobs. This is also the case for Zeppelin and Spark-notebook. It's not a real blocker, but it would be nice if we could find a way to allow other jobs to run on that node in parallel... Not sure if this is possible..

File: lib/charms/spark.py [hunk 1]
```diff
@@ -1,30 +1,26 @@
...
-from charmhelpers.core import unitdata
 from charmhelpers.fetch.archiveurl import ArchiveUrlFetchHandler
-from jujubigdata import utils
-
 
 # Main Spark class for callbacks
 class Spark(object):
     def __init__(self, dist_config):
         self.dist_config = dist_config
+        print(self.dist_config)
```
Comment (ktsakalozos) [inline lib/charms/spark.py]: Lets log instead of print

File: lib/charms/spark.py [hunk 2]
```diff
@@ -34,15 +30,16 @@ def install(self, force=False):
...
+            '{}/lib/spark-assembly-*.jar'.format(self.dist_config.path('spark'))
+        )[0]
         utils.run_as('hdfs', 'hdfs', 'dfs', '-mkdir', '-p',
                      '/user/ubuntu/share/lib')
         try:
             utils.run_as('hdfs', 'hdfs', 'dfs', '-put', spark_assembly_jar,
                          '/user/ubuntu/share/lib/spark-assembly.jar')
         except CalledProcessError:
-            print ("File exists")
+            print("File exists")
```
Comment (ktsakalozos) [inline lib/charms/spark.py]: Same here, lets log

File: reactive/spark.py [hunk 3]
```diff
@@ -53,38 +54,97 @@ def install_spark(hadoop):
...
 
-@when('spark.installed', 'hadoop.ready')
+@when('hadoop.ready', 'spark.installed')
+@when_not('livy.installed')
+def install_livy(hadoop):
+    dist = DistConfig(data=layer.options('livy'))
+    livy = Livy(dist)
+    if livy.verify_resources():
+        hookenv.status_set('maintenance', 'Installing Livy REST server')
+        dist.add_users()
```
Comment (ktsakalozos) [inline reactive/spark.py]: I prefer adding uses/groups/directories to be handled inside the Livy/Spark class. The reason is that the Livy/Spark objects consume the DistConfig and know if they need users/dirs/groups

File: lib/charms/spark.py (hunk 1 above)
Comment (galgalesh) [inline lib/charms/spark.py]: Thanks for pointing this out, this should be removed altogether, I added it for debugging.

File: reactive/spark.py (hunk 3 above)
Comment (galgalesh) [inline reactive/spark.py]: Good idea!

File: layer.yaml [hunk 4]
```diff
@@ -1,31 +1,59 @@
-includes: ['layer:hadoop-client', 'interface:spark', 'interface:benchmark']
+includes:
+- layer:hadoop-client
+- interface:spark
+- interface:benchmark
 options:
-  hadoop-client:
+  apache-spark:
```
Comment (johnsca) [inline layer.yaml]: This isn't how layer options work.  The layer name here is the layer for which you are setting the options, or, in other words, the layer that defines the options you want to set.  In this case, that's `hadoop-client` and setting options for any other layer (`apache-spark`, in this case) will cause them to be ignored (it should probably complain during build instead).

File: layer.yaml [hunk 5]
```diff
@@ -1,31 +1,59 @@
...
+    ports:
+      livy:
+        port: 8998
+        exposed_on: spark
+    groups: []
+  basic:
+    use_venv: false
+    packages: []
+    include_system_packages: false
+is: apache-spark
```
Comment (johnsca) [inline layer.yaml]: The `is` key is an artifact of the build process and shouldn't be set by layers.  (Though, I doubt it hurts anything.)

File: lib/charms/spark.py [hunk 6]
```diff
@@ -183,37 +179,31 @@ def configure(self):
...
-                    r'^SPARK_EXECUTOR_MEMORY.*': '# SPARK_EXECUTOR_MEMORY (use value from environment)',
-                    })
-
-                unitdata.kv().set('spark_bench.installed', True)
-                unitdata.kv().flush(True)
+            if utils.cpu_arch() == 'ppc64le':
+                sb_url = hookenv.config()['spark_bench_ppc64le']
+            else:
+                # TODO: may need more arch cases (go with x86 sb for now)
+                sb_url = hookenv.config()['spark_bench_x86_64']
```
Comment (johnsca) [inline lib/charms/spark.py]: Probably worth changing this at this point to:

```python
sb_url = hookenv.config()['spark_bench_%s' % utils.cpu_arch()]
```

File: reactive/spark.py [hunk 7]
```diff
@@ -32,59 +35,113 @@ def report_blocked():
...
+    livy.open_ports()
+    set_state('livy.started')
+    hookenv.status_set('active', 'Ready')
+
+
+@when('spark.installed', 'livy.installed', 'hadoop.ready', 'spark.started', 'livy.started')
+def reconfigure_spark(hadoop):# pylint: disable=w0613
+    config = hookenv.config()
+    if not data_changed('configuration', config):
+        return
```
Comment (johnsca) [inline reactive/spark.py]: I'd like to suggest adding `@when('config.changed')` instead if we're not watching a specific subset of config.

File: lib/charms/spark.py [hunk 8]
```diff
@@ -19,12 +18,13 @@ def __init__(self, dist_config):
...
 
-    def install(self, force=False):
-        if not force and self.is_installed():
-            return
+    def install(self):
+        '''
+        install Spark and add dependencies in dist-config
+        '''
+        self.dist_config.add_dirs()
+        self.dist_config.add_packages()
```
Comment (johnsca) [inline lib/charms/spark.py]: I'm realizing that the `hadoop-client` layer should either act on the options it defines (in which case it would need to expose another state to indicate when that was done), or not define the options at all and leave it up to the client charm.

I'm not sure which would be better / more user-friendly.

File: layer.yaml (hunk 4 above)
Comment (galgalesh) [inline layer.yaml]: Hi Cory

Thanks for your thorough feedback. Can you explain a bit more what the idea
behind the options is? I thought this was just an upstreamed replacement
for dist.yaml. I'm not sure why these options schould be passed to the
hadoop-client. I am in fact setting them for and using them in the spark
layer. I'm not sure why these should be passed to another layer.

Op maandag 29 februari 2016 heeft Cory Johns <notifications@github.com> het
volgende geschreven:
> In layer.yaml:
>
>> options:
>> - hadoop-client:
>> + apache-spark:
>
> This isn't how layer options work. The layer name here is the layer for
which you are setting the options, or, in other words, the layer that
defines the options you want to set. In this case, that's hadoop-client and
setting options for any other layer (apache-spark, in this case) will cause
them to be ignored (it should probably complain during build instead).
>
> —
> Reply to this email directly or view it on GitHub.<
https://ci5.googleusercontent.com/proxy/C4LZzv7PlLgC2rWZmE6Z4GHtk3I0aJtZCCFZMzFVz--8DYfxjaezvrp7hT5fyqrIL3mP3FFp3hYDv_6KWdNteyXOQIz8lx2aT93guJk5NDKbd4zam7Si1Nv5V4FyNXk_wamVDnhWc1nLCdJC-8w3KjbSfEpKDQ=s0-d-e1-ft#https://github.com/notifications/beacon/ABbH9XVzH5ToQCxj6VcXX4qHzQimE-tGks5poz44gaJpZM4Hkmc3.gif
>

Comment (johnsca) [inline layer.yaml]: I completely missed that you manually pull out the layer options in `reactive/spark.py`, so let me rephrase that: this isn't how layer options are *supposed* to work.  :)  Specifically, layer options are intended to be a method for charm layers to communicate to base layers and provide them data and otherwise modify their behavior.

There's no reason that this layer can't define its own options that it uses, but they should still be namespaced under `apache-spark` and should use the `defines` section to describe how they should be validated and defaults to give them values.

That said, since the Kafka charm layer also uses these options from `hadoop-client`, my inclination is to move the implementation for them to that layer, which unfortunately would mean removing the separation (livy vs spark) that you added, which does seem nice to have.

Comment (johnsca) [inline layer.yaml]: I didn't see your reply while I was drafting the previous, but I think it addresses some of your questions.  However, to address explicitly why it would belong in `hadoop-client`, my thought in structuring it that way was that it was common functionality that could be provided by the base layer (i.e., so that not every client layer had to remember to call `dist_config.add_users()`, etc.), and layer options seemed the most natural way to send that data from the charm layer to the base layer.

Layer options was created as a way to provide a well defined, build-time validated way to pass data from charm layers to base layers, and to do away with the arbitrary yaml files like `dist.yaml` and `apache.yaml`.

Comment (galgalesh) [inline layer.yaml]: Thanks for this explanation. I understand where you'd like to go to. I'll
look into the `defines` section. I really like the idea of handing off
common work to lower layers. I wouldn't put the handeling of `dist` in the
hadoop-client layer however. This is functionality that I'd like to use in
other Charms that are not Hadoop clients. It might be better to put this in
a 'dist' layer (or a more descriptive name).

I think the separation between Spark and Livy is very important. First of
all because, i think, it makes code a lot cleaner. Second of all, because
we'll need an approach that can scale. There are a number of similar
projects out there and if the ecosystem starts to grow, so will the number
of additional api's and services. We'll need to be able to handle that
scale, and starting with clear separation between [optional] sub-services
 seems like a step in the right direction.

I'll create a PR to address your concerns. I'd like to keep the separation
of Livy vs Spark vs hadoop-client for now. There isn't much code
duplication anyway because most of the hard work is done by the jujubigdata
module. Although this is something that should be discussed further.



2016-02-29 22:24 GMT+01:00 Cory Johns <notifications@github.com>:

**Comment by arosales** [comment]
@galgalesh 

I don't know where you currently are with Livy, but I wanted to chime in on this open PR and suggest a way forward.

The current PR is to add Livy directly into Spark.  At the time of the PR we only had one Spark, but now we have 2 Spark charms for Apache upstream and BigTop. Most recent work has been put into the Bigtop charms. If we wanted to get Livy there we would also need to investigate modifying the bigtop-spark layer. This prompted me to think if we should consider Livy as a subordinate that could be deployed onto Apache, Bigtop, Cloudera, or others without modifications to the spark layer.

Thus, I wanted to get your thoughts on Livy as either its own layer that gets built into charms (ie not part of the spark layer itself), or as a subordinate?

Thoughts?

**Comment by merlijn-sebrechts** [comment]
I have no issue with Livy as a subordinate Charm. I propose to close this PR, Livy isn't a priority for us (tengu) at the moment so it will probably be a while until I reboot this work as a subordinate Charm.
